cmd/endpoint.go

```
@@ -0,0 +1,145 @@
+	UnknownEndpointType EndpointType = 0
+
+	// FSEndpointType - FS endpoint type enum.
+	FSEndpointType = 1
+
+	// XLEndpointType - XL endpoint type enum.
+	XLEndpointType = 2
+)
+
+// Endpoint - any type of endpoint.
+type Endpoint struct {
+	Index      int
+	Value      string
+	IsDisabled bool
+	URL        *url.URL
+}
+
+// Type - returns type of endpoint.
+func (endpoint Endpoint) Type() EndpointType {
+	if IsEmptyPath(endpoint.Value) {
+		return UnknownEndpointType
+	}
+
+	if endpoint.URL != nil {
+		return XLEndpointType
+	}
+
+	return FSEndpointType
+}
+
+// NewEndpoint - returns new endpoint based on given arguments.
+func NewEndpoint(arg string, index int) (Endpoint, error) {
+	if IsEmptyPath(arg) {
+		return Endpoint{}, fmt.Errorf("Empty or root endpoint is not supported")
+	}
+
+	u, err := url.Parse(arg)
+	if err == nil && u.Host != "" {
+		// May be XL type endpoint arg.
+		host, port, err := splitHostPort(u.Host)
+		if !(err == nil && host != "" && (u.Scheme == "http" || u.Scheme == "https") &&
+			u.Opaque == "" && u.ForceQuery == false && u.RawQuery == "" && u.Fragment == "") {
+			return Endpoint{}, fmt.Errorf("Unknown endpoint format")
+		}
+
+		if IsEmptyPath(u.Path) {
+			return Endpoint{}, fmt.Errorf("Empty or root path is not supported in URL endpoint")
+		}
+
+		u.Host = host + ":" + port
+		arg = u.String()
+	} else {
+		u = nil
+	}
+
+	return Endpoint{
+		Index: index,
+		Value: arg,
+		URL:   u,
+	}, nil
+}
+
+// EndpointList - endpoint list hashed with endpoint's value
+type EndpointList map[string]Endpoint
+
+// NewEndpointList - returns new endpoint list based on input args.
+func NewEndpointList(args ...string) (endpoints EndpointList, err error) {
+	// Check whether given args contain duplicates.
+	if uniqueArgs := set.CreateStringSet(args...); len(uniqueArgs) != len(args) {
+		return nil, fmt.Errorf("duplicate endpoints found")
+	}
+
+	// Check whether no. of args are valid for XL distribution.
+	if !IsValidDistribution(len(args)) {
```

IsValidDistribution() accepts only one argument, which means NewEndpointList works only with XL/Distributed (not obvious), can we make this check outside this function, it will be less confusing.

---

I couldn't able to get you.

NewEndpointList() is applicable only to XL or Distribute setup.  The arguments can be multiples to such type of setup and we would need to fail if `len(args)` are unsupported.

Could you tell me why `len(args) = 1` is a valid case here?

---

> NewEndpointList() is applicable only to XL or Distribute setup.

My point is that NewEndpointList() name doesn't indicate that this is about XL or Distributed. And, moving IsValidDistribution() outside NewEndpointList() will make the latter work with FS too (in case if we need something like this later)

---

I think you are talking about FS type of `Setup` which is already taken care in `FSSetup`.

As you see in `NewEndpointList()`, any type of endpoints can be added but they are all have to be same type ie either `FSEndpointType` or `XLEndpointType`

If you are taking about FS/XL/Distribute setup, they are defined in `setup.go` where they use `NewEndpointList()`.

The idea of keeping/doing check of `IsValidDistribution()` at `NewEndpointList()` is `EndpointList` can not have a wrong set of endpoints.

Do you say `EndpointList` could have any length of endpoints where `XLSetup` and `DistributeSetup` should take care of distribution check?

---

> Do you say EndpointList could have any length of endpoints where XLSetup and DistributeSetup should take care of distribution check?

Yes. Or you can rename NewEndpointList() to NewXLEndpointList(), my concern is mostly about naming.

---

@vadmeste 

> Yes. Or you can rename NewEndpointList() to NewXLEndpointList(), my concern is mostly about naming.

I like the renaming approach, but one thing about `NewXLEndpointList()` is something not same as golang's style like a `<T>` and `New<T>()` for creating new `<T>`.

Let me know whether I go ahead of renaming.